### ISSUE TITLE
Support Shadow DOM Styling

### ISSUE THREAD

**Issue by nitipit** [keenlycode/adapter#8]
The answer could be here : https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM#constructable_stylesheets
By applying **CSS** Programmatically], using [CSSStyleSheet](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet) object and attaching it to the shadow root.

**Comment by nitipit** [comment]
One thing that needs support is ShadowDOM. Over here:

https://github.com/keenlycode/adapter/blob/9419c46ad62e3d9b30951a12d5ba5445e38a4236/src/adapter.ts#L30

addStyle will always add the style to document, but the problem is that if an element is inside a ShadowRoot then it won't style.

The non-static addStyle does not accept a parentNode argument, so not possible to solve using that. But _addStyle could be used directly.
Ideally user would do something like MyClass.tagStyle(...) and when they append the new MyClass() inside of some ShadowRoot, it would just work.

Lume has ShadowDOM, f.e. here's the <lume-camera-rig> element's template which gets rendered in its ShadowRoot:

https://github.com/lume/lume/blob/83560218df78376162cb267f92cc2d87dd44a5b8/src/cameras/CameraRig.ts#L279-L311
In @lume/element, we apply static css or css (non-static) like this in connectedCallback:

https://github.com/lume/element/blob/aaf32e77c414ee709d24aa30fa367baf634b58d0/src/LumeElement.ts#L251

and you can see there depending on if there is a ShadowRoot or not (uses this.getRootNode() to get the element's root node when there is not a ShadowRoot)

Perhaps  Adapter could do similar to that, and it can grab whatever static style at that time and either place it in the element's own ShadowRoot, or otherwise in the nearest root node (which could be some other ShadowRoot)
Joe Pea — 12/05/2023 7:32 AM
.
One thing to keep in mind when implementing this is that this.shadowRoot might already exist, so you you might want to detect that. In Lume's particular case, if this.hasShadow is true, there is a shadow root (whether open or closed).

You can see here that LumeElement overrides attachShadow so that no matter if it is open or closed, it can have a reference:

https://github.com/lume/element/blob/aaf32e77c414ee709d24aa30fa367baf634b58d0/src/LumeElement.ts#L243-L246

The question is, how to make Adapter robust so that it works with any element that may have an open or closed ShadowRoot.

Perhaps you can just see if this.shadowRoot is not null in connectedCallback. If you have a root, use it (describe in the docs that Adapter appends a style by default so people know). Provide an option to disable auto-append style just in case.

Then, also provide a method like applyStyle() (no args) that a custom element author can call to apply styles manually in case (1) Adapter is not able to do it automatically in case of a closed root or (2) the user opted out of the automatic behavior.

Sample user class

```js
class MyEl extends Adapter(LitElement) {
  autoApplyStyle = false

  connectedCallback() {
    super.connectedCallback()
    // ... do some stuff with ShadowRoot ...
    this.applyStyle() // finally append Adapter style.
  }
}
```

**Comment by nitipit** [comment]
Scoped Element might do the trick

**Comment by nitipit** [comment]
This has been implemented already in v2.4.0
